### PR TITLE
Space Array Game

### PR DESCRIPTION
**Pull request by Sakshi1007**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

yes

### Description

This is a game problem very famous in competitive programming. I have added the code of this problem with problem description and also some test cases. 

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

Resolves: #3264

### CODE REVIEW

**Review by iamrajiv — CHANGES_REQUESTED** [review]
@Sakshi1007 Add space and time complexity. Also, add the problem link.

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Why does this problem look similar to [this](https://www.hackerearth.com/docs/wiki/developers/solution-guide/) ?

**Comment by Sakshi1007** [comment]
Please give Review for above problem.
Changes requested are done.

**Review by iamrajiv — APPROVED** [review]

**Comment by Sakshi1007** [comment]
@HarshCasper  plz review my PR

**Comment by Sakshi1007** [comment]
@HarshCasper, 
@cOsmOvoid , @adarshvulli plz review my PR

File: C-Plus-Plus/cp/Space_Array.cpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+if the first player wins or "Second"
+if the second player wins (without quotes).*/
+
+/*
+Constraints
+1=T=2·104
+1=N=2·105
+The sum of N over all test cases doesn't exceed 2·105
+1=ai=N for each valid i
+*/
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Space_Array.cpp]: No need of explaining the constraints

File: C-Plus-Plus/cp/Space_Array.cpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+/*
+Input:
+The first line of the input contains a single integer T denoting the number of test cases.
+The description of T test cases follows.
+The first line of each test case contains a single integer N.
+The second line contains N space-separated integers a1,a2,…,aN.
+Output:
+For each test case, print a single line containing the string "First"
+if the first player wins or "Second"
+if the second player wins (without quotes).*/
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Space_Array.cpp]: No need for this as well

File: C-Plus-Plus/cp/Space_Array.cpp [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+The rules of the game are as follows:
+
+1-Initially, a sequence a1,a2,…,aN is given.
+2-The players alternate turns.
+3-In each turn, the current player must choose
+an index i and increment ai, i.e. change ai to ai+1.
+4-Afterwards, if there is no permutation p1,p2,…,pN of the
+integers 1 through N such that ai=pi holds for each valid
+ i, the current player loses.
+5-Otherwise, the game continues with the next turn.
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Space_Array.cpp]: Keep it more standard

File: C-Plus-Plus/cp/Space_Array.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+if array at that index is greater than the array then Second Player will be the winner.
+if not the we will take the sum of difference of index and array at that index.
+sum will be odd then first will win otherwise second will win
+*/
+
+/*Time complexity: 0.06 sec using above algo, "MAX time complexity can be 1 sec "
+Memory used: 6.5 MB uisng above algo, MAx space complexity for his problem is: 50000 Bytes"
+*/
+
+/* PROBLEM LINK: https://www.codechef.com/MARCH21C/problems/SPACEARR
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Space_Array.cpp]: Shift it to the top

File: C-Plus-Plus/cp/Space_Array.cpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+Second
+*/
+
+/*
+Explanation:
+if array at that index is greater than the array then Second Player will be the winner.
+if not the we will take the sum of difference of index and array at that index.
+sum will be odd then first will win otherwise second will win
+*/
+
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Space_Array.cpp]: Try to embed the explanation in the code itself

File: C-Plus-Plus/cp/Space_Array.cpp [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+1=T=2·104
+1=N=2·105
+The sum of N over all test cases doesn't exceed 2·105
+1=ai=N for each valid i
+*/
+
+// Code
+#include<bits/stdc++.h>
+using namespace std;
+int main() {
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Space_Array.cpp]: Use a different function for the main logic of the code. Use `main()` function for the driver code

**Review by HarshCasper — CHANGES_REQUESTED** [review]

File: C-Plus-Plus/cp/Space_Array.cpp (hunk 1 above)
Comment (Sakshi1007) [reply C-Plus-Plus/cp/Space_Array.cpp]: Sure I will remove

File: C-Plus-Plus/cp/Space_Array.cpp (hunk 2 above)
Comment (Sakshi1007) [reply C-Plus-Plus/cp/Space_Array.cpp]: Ok

File: C-Plus-Plus/cp/Space_Array.cpp (hunk 4 above)
Comment (Sakshi1007) [reply C-Plus-Plus/cp/Space_Array.cpp]: Ok i will.

File: C-Plus-Plus/cp/Space_Array.cpp (hunk 5 above)
Comment (Sakshi1007) [reply C-Plus-Plus/cp/Space_Array.cpp]: Okk sure. Thank u

**Comment by Sakshi1007** [comment]
@HarshCasper plz review my PR

File: C-Plus-Plus/cp/Space_Array.cpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+4-Afterwards, if there is no permutation p1,p2,…,pN of the
+integers 1 through N such that ai=pi holds for each valid
+ i, the current player loses.
+5-Otherwise, the game continues with the next turn.
+*/
+
+// Code
+#include<bits/stdc++.h>
+using namespace std;
+int main() {
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Space_Array.cpp]: Use a different function for the main logic of the code. Use main() function for the driver code

File: C-Plus-Plus/cp/Space_Array.cpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+/*
+The rules of the game are as follows:
+1-Initially, a sequence a1,a2,…,aN is given.
+2-The players alternate turns.
+3-In each turn, the current player must choose
+an index i and increment ai, i.e. change ai to ai+1.
+4-Afterwards, if there is no permutation p1,p2,…,pN of the
+integers 1 through N such that ai=pi holds for each valid
+ i, the current player loses.
+5-Otherwise, the game continues with the next turn.
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Space_Array.cpp]: Keep it more standard

File: C-Plus-Plus/cp/Space_Array.cpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+		}
+		if (c == 1)                             //c is 1 i.e. arrayy is greater than element of array
+			cout << "Second" << endl;        //second player wins.
+		else {
+			if (sum % 2 == 0)                  // if sum is even.
+				cout << "Second" << endl;      // second is winner.
+			else
+				cout << "First" << endl;       // otherwise first will win.
+		}
+	}
```
Comment (HarshCasper) [inline C-Plus-Plus/cp/Space_Array.cpp]: Keep comments above the line of code not beside it

**Review by HarshCasper — CHANGES_REQUESTED** [review]

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Don't use comments beside the line of code. Use them above the line of code

**Comment by Sakshi1007** [comment]
Done.

**Review by HarshCasper — APPROVED** [review]

**Pull request merged**